ably-ios/ARTBaseMessage.m

```diff
@@ -40,20 +40,26 @@ - (instancetype)messageWithData:(id)data encoding:(NSString *)encoding {
     return message;
 }
 
-- (ARTStatus *)decodeWithEncoder:(ARTDataEncoder*)encoder output:(id *)output {
+- (id)decodeWithEncoder:(ARTDataEncoder*)encoder error:(NSError **)error {
     ARTDataEncoderOutput *decoded = [encoder decode:self.data encoding:self.encoding];
-    *output = [self copy];
-    ((ARTBaseMessage *)*output).data = decoded.data;
-    ((ARTBaseMessage *)*output).encoding = decoded.encoding;
-    return decoded.status;
+    if (decoded.status.state != ARTStateOk && error) {
+        *error = [NSError errorWithDomain:ARTAblyErrorDomain code:0 userInfo:@{@"reason": @"decoding failed"}];
```

Comment:
Please use `NSLocalizedDescriptionKey` or `NSLocalizedFailureReasonErrorKey`. It's a convention throughout Cocoa.

 - NSLocalizedDescriptionKey: A localized description of the error.
 - NSLocalizedFailureReasonErrorKey: A localized explanation of the reason for the error.

E.g.:
```
NSDictionary *userInfo = @{ NSLocalizedDescriptionKey: @"Operation was unsuccessful." };
```